src/stan/model/indexing/rvalue.hpp

```diff
@@ -18,28 +18,44 @@ namespace model {
 // all indexing from 1
 
 /**
- * Return the result of indexing a specified value with
- * a nil index list, which just returns the value.
+ * Return the result of indexing a specified scalar type with
+ * a nil index list, which just returns the scalar.
  *
  * Types:  T[] : T
  *
  * @tparam T Scalar type.
  * @param[in] c Value to index.
  * @return Input value.
  */
-template <typename T>
-inline T rvalue(const T& c, const nil_index_list& /*idx*/,
-                const char* /*name*/ = "", int /*depth*/ = 0) {
+template <typename T, typename = require_stan_scalar_t<T>>
+inline T rvalue(T&& c, const nil_index_list& /*idx*/, const char* /*name*/ = "",
+                int /*depth*/ = 0) {
   return c;
 }
 
+/**
+ * Return the result of indexing a specified non-scalar value with
+ * a nil index list, which just returns the value.
+ *
+ * Types:  T[] : T
+ *
+ * @tparam T Scalar type.
+ * @param[in] c Value to index.
+ * @return Input value.
+ */
+template <typename T, typename = require_not_stan_scalar_t<T>>
+inline auto&& rvalue(T&& c, const nil_index_list& /*idx*/,
```

Comment:
`auto` is deduced as a value so this will always return a rvalue reference. `decltype(auto)` is instead deduced as appropriate reference type (no `&&` is needed).
```suggestion
inline decltype(auto) rvalue(T&& c, const nil_index_list& /*idx*/,
```

Reply:
I'm cool with changing this to `declar_type(auto)`, from reading the stackoverflow answer below I though that `auto&&` would preserve the returns value type and cv qualifier?

https://stackoverflow.com/a/13242177/2269255

Reply:
Right, after some playing around I found out It works the same. While I feel like `decltype(auto)` might be more common in such situations, I will leave it up to you to decide which one to use in your PR.

Reply:
I think it's more common to see `decltype(auto)` as the return type so let's leave it as is